# Introducción

Para esta práctica se realizarán los ejercicios del 6.9 al 6.13 del libro brindado en la materia, empleando las transformaciones vistas en clase.

In [2]:
import numpy as np
import matplotlib.pyplot as plt

# Problema 6.9
### Alinéese el vector **V = I + J + K**  con el vector **K**.

$$
A_v = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & -\frac{ab}{\lambda|\vec{V}|} & -\frac{ac}{\lambda|\vec{V}|} & 0 \\
    0 & \frac{c}{\lambda} & -\frac{b}{\lambda} & 0 \\
    \frac{a}{|\vec{V}|} & \frac{b}{|\vec{V}|} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
; \quad \vec{V} = 
\begin{pmatrix}
    1 \\
    1 \\
    1 \\
    1
\end{pmatrix}
$$

Donde:

$$
\begin{align*}
a &= 1 \\
b &= 1 \\
c &= 1 \\
\lambda &= \sqrt{b^2 + c^2} = \sqrt{1^2 + 1^2} = \sqrt{2} \\
|\vec{V}| &= \sqrt{a^2 + b^2 + c^2} = \sqrt{1^2 + 1^2 + 1^2} = \sqrt{3}
\end{align*}
$$

In [3]:
# Valores dados
a = 1
b = 1
c = 1
lambda_ = np.sqrt(b**2 + c**2)
v_magnitude = np.sqrt(a**2 + b**2 + c**2)
V = np.array([1, 1, 1, 1])

Obtención de la matriz de alineación.

$$
A_v = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & -\frac{ab}{\lambda|\vec{V}|} & -\frac{ac}{\lambda|\vec{V}|} & 0 \\
    0 & \frac{c}{\lambda} & -\frac{b}{\lambda} & 0 \\
    \frac{a}{|\vec{V}|} & \frac{b}{|\vec{V}|} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
= 
\begin{pmatrix}
    \frac{\sqrt{2}}{\sqrt{3}} & -\frac{1\cdot1}{\sqrt{2}\cdot\sqrt{3}} & -\frac{1\cdot1}{\sqrt{2}\cdot\sqrt{3}} & 0 \\
    0 & \frac{1}{\sqrt{2}} & -\frac{1}{\sqrt{2}} & 0 \\
    \frac{1}{\sqrt{3}} & \frac{1}{\sqrt{3}} & \frac{1}{\sqrt{3}} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
= 
\begin{pmatrix}
    \frac{\sqrt{6}}{3} & -\frac{\sqrt{6}}{6} & -\frac{\sqrt{6}}{6} & 0 \\
    0 & \frac{\sqrt{2}}{2} & -\frac{\sqrt{2}}{2} & 0 \\
    \frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
$$

In [4]:
# Matriz de alineación
A_v = np.array([
    [lambda_ / v_magnitude, -a*b / (lambda_ * v_magnitude), -a*c / (lambda_ * v_magnitude), 0],
    [0, c / lambda_, -b / lambda_, 0],
    [a / v_magnitude, b / v_magnitude, c / v_magnitude, 0],
    [0, 0, 0, 1]
])

Realizamos la operación $V' = A_v \cdot V$.

$$
V' = \begin{pmatrix}
    \frac{\sqrt{6}}{3} & -\frac{\sqrt{6}}{6} & -\frac{\sqrt{6}}{6} & 0 \\
    0 & \frac{\sqrt{2}}{2} & -\frac{\sqrt{2}}{2} & 0 \\
    \frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix} \begin{pmatrix}
    1 \\
    1 \\
    1 \\
    1
\end{pmatrix} = \begin{pmatrix}
    0 \\
    0 \\
    \sqrt{3} \\
    1
\end{pmatrix} \approx \begin{pmatrix}
    0 \\
    0 \\
    1.732 \\
    1
\end{pmatrix}
$$

In [5]:
# Calculamos V'
V_prime = np.dot(A_v, V)

V_prime

array([2.22044605e-16, 0.00000000e+00, 1.73205081e+00, 1.00000000e+00])


Dado que $2.22044605 \times 10^{-16}$ es efectivamente 0 dentro de la precisión numérica el vector **V** alineado con el vector **K** es **V' = (0, 0, 1.732, 1)**.

# Problema 6.10
### Obténgase una transformación que alinea al vector **V = I + J + K** con el vector **N = 2I - J - K**.

$$
A_v = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & -\frac{ab}{\lambda|\vec{V}|} & -\frac{ac}{\lambda|\vec{V}|} & 0 \\
    0 & \frac{c}{\lambda} & -\frac{b}{\lambda} & 0 \\
    \frac{a}{|\vec{V}|} & \frac{b}{|\vec{V}|} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
; \quad \vec{V} = 
\begin{pmatrix}
    1 \\
    1 \\
    1 \\
    1
\end{pmatrix}
; \quad \vec{N} = 
\begin{pmatrix}
    2 \\
    -1 \\
    -1 \\
    1
\end{pmatrix}
$$

Donde:

$$
\begin{align*}
a &= 2 \\
b &= -1 \\
c &= -1 \\
\lambda &= \sqrt{b^2 + c^2} = \sqrt{(-1)^2 + (-1)^2} = \sqrt{1 + 1} = \sqrt{2} \\
|\vec{V}| &= \sqrt{a^2 + b^2 + c^2} = \sqrt{2^2 + (-1)^2 + (-1)^2} = \sqrt{4 + 1 + 1} =\sqrt{6}
\end{align*}
$$

In [9]:
# Valores dados
a = 2
b = -1
c = -1
lambda_ = np.sqrt(b**2 + c**2)
v_magnitude = np.sqrt(a**2 + b**2 + c**2)
V = np.array([1, 1, 1, 1])

Obtención de la matriz de alineación.

$$
A_v = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & -\frac{ab}{\lambda|\vec{V}|} & -\frac{ac}{\lambda|\vec{V}|} & 0 \\
    0 & \frac{c}{\lambda} & -\frac{b}{\lambda} & 0 \\
    \frac{a}{|\vec{V}|} & \frac{b}{|\vec{V}|} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
= 
\begin{pmatrix}
    \frac{\sqrt{2}}{\sqrt{6}} & \frac{-2\cdot-1}{\sqrt{2}\cdot\sqrt{6}} & \frac{-2\cdot-1}{\sqrt{2}\cdot\sqrt{6}} & 0 \\
    0 & \frac{-1}{\sqrt{2}} & -\frac{-(1)}{\sqrt{2}} & 0 \\
    \frac{2}{\sqrt{6}} & \frac{-1}{\sqrt{6}} & \frac{-1}{\sqrt{6}} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
= 
\begin{pmatrix}
    \sqrt{3} & \frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} & 0 \\
    0 & -\frac{\sqrt{2}}{2} & \frac{\sqrt{6}}{6} & 0 \\
    \frac{\sqrt{6}}{3} & -\frac{\sqrt{6}}{6} & -\frac{\sqrt{6}}{6} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
$$

In [10]:
# Matriz de alineación
A_v = np.array([
    [lambda_ / v_magnitude, -a*b / (lambda_ * v_magnitude), -a*c / (lambda_ * v_magnitude), 0],
    [0, c / lambda_, -b / lambda_, 0],
    [a / v_magnitude, b / v_magnitude, c / v_magnitude, 0],
    [0, 0, 0, 1]
])

Realizamos la operación $V' = A_v \cdot V$.

$$
V' = \begin{pmatrix}
    \sqrt{3} & \frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} & 0 \\
    0 & -\frac{\sqrt{2}}{2} & \frac{\sqrt{6}}{6} & 0 \\
    \frac{\sqrt{6}}{3} & -\frac{\sqrt{6}}{6} & -\frac{\sqrt{6}}{6} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix} \begin{pmatrix}
    1 \\
    1 \\
    1 \\
    1
\end{pmatrix} = \begin{pmatrix}
    \frac{2}{\sqrt{3}}+\sqrt{3}\\
    \frac{1}{\sqrt{6}}-\frac{1}{\sqrt{2}}\\
    0 \\
    1
\end{pmatrix} \approx \begin{pmatrix}
    2.88675 \\
    -0.298858 \\
    0 \\
    1
\end{pmatrix}
$$

La transformación que alinea al vector **V** con el vector **N** es **V' = (2.88675, -0.298858, 0, 1)**.

# Problema 6.11 

### Muéstrese que la transformación de alineamiento stisface la relación $A_v^{-1} = A_v^T$.

$$
A_v = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & -\frac{ab}{\lambda|\vec{V}|} & -\frac{ac}{\lambda|\vec{V}|} & 0 \\
    0 & \frac{c}{\lambda} & -\frac{b}{\lambda} & 0 \\
    \frac{a}{|\vec{V}|} & \frac{b}{|\vec{V}|} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
$$

$$
A_v^{-1} = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & 0 & \frac{a}{|\vec{V}|} & 0 \\
    -\frac{ab}{\lambda|\vec{V}|} & \frac{c}{\lambda} & \frac{b}{|\vec{V}|} & 0 \\
    -\frac{ac}{\lambda|\vec{V}|} & -\frac{b}{\lambda} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
$$

$$
A_v^T = A_v^{-1} = 
\begin{pmatrix}
    \frac{\lambda}{|\vec{V}|} & 0 & \frac{a}{|\vec{V}|} & 0 \\
    -\frac{ab}{\lambda|\vec{V}|} & \frac{c}{\lambda} & \frac{b}{|\vec{V}|} & 0 \\
    -\frac{ac}{\lambda|\vec{V}|} & -\frac{b}{\lambda} & \frac{c}{|\vec{V}|} & 0 \\
    0 & 0 & 0 & 1
\end{pmatrix}
$$

Por lo tanto, la transformación de alineamiento satisface la relación $A_v^{-1} = A_v^T$.

# Problema 6.12
### Muéstrese que la matriz de alineamiento $A_{V,N}$ es equivalente a una rotación de $\theta°$ con respecto a un eje que tiene la dirección del vector $V \times N$, y pasa por el origen. Aquí, $\theta$ es el ángulo entre los vectores $V$ y $N$.

El posicionamiento de los vectores $\mathbf{V}$ y $\mathbf{N}$ en el origen nos permite observar que el producto cruz $\mathbf{V} \times \mathbf{N}$ resulta ser perpendicular a ambos $\mathbf{V}$ y $\mathbf{N}$. Al considerar $\theta$ como el ángulo comprendido entre $\mathbf{V}$ y $\mathbf{N}$, se deduce que una rotación de $\theta^\circ$ en torno al eje formado por $\mathbf{A} \times \mathbf{N}$, que intercepta el origen, produce la alineación de $\mathbf{V}$ con $\mathbf{N}$. Esto sustenta la afirmación de que $\mathbf{A}_{V,N} = \mathbf{R}_{\theta, \mathbf{j}}$.

Para el producto cruz se tiene que:

$$
\mathbf{V} \times \mathbf{N} = \det
\begin{pmatrix}
\mathbf{I} & \mathbf{J} & \mathbf{K} \\
a & b & c \\
n_1 & n_2 & n_3
\end{pmatrix}
= (b n_3 - c n_2) \mathbf{I} + (c n_1 - a n_3) \mathbf{J} + (a n_2 - b n_1) \mathbf{K}
$$

El producto escalar se expresa como:

$$
(\mathbf{a} \quad \mathbf{b} \quad \mathbf{c}) \cdot (n_1 \quad n_2 \quad n_3) = a n_1 + b n_2 + c n_3
$$

La relación entre el producto escalar de $\mathbf{V}$ y $\mathbf{N}$, y el ángulo $\theta$ entre ellos, se define por:

$$
\mathbf{V} \cdot \mathbf{N} = |\mathbf{V}| \cdot |\mathbf{N}| \cos \theta
$$

Por lo tanto, el ángulo $\theta$ se determina mediante:

$$
\theta = \arccos \left( \frac{\mathbf{V} \cdot \mathbf{N}}{|\mathbf{V}| \cdot |\mathbf{N}|} \right) = \arccos \left( \frac{a n_1 + b n_2 + c n_3}{\sqrt{a^2 + b^2 + c^2} \cdot \sqrt{n_1^2 + n_2^2 + n_3^2}} \right)
$$

Estableciendo así la igualdad $\mathbf{A}_{V,N} = \mathbf{R}_{\theta, \mathbf{j}}$.


# Problema 6.13 

### ¿Cómo puede definirse la puesta en escala con respecto a un punto $P_0(x_0, y_0, z_0)$  en términos de puesta en escala con respecto al origen?

Inicialmente, se requiere posicionar el punto de interés en el origen. Posteriormente, se ejecuta la operación de escalamiento respecto al origen y finalmente se reposiciona el punto a su ubicación original.

$$
S_{s_x,s_y,s_z,P_0} = (T_{-P_0})^{-1} \cdot S \cdot T_{-P_0}
$$

Donde la transformación compuesta para el escalamiento con respecto a un punto $P_0(x_0, y_0, z_0)$, reubicándolo al origen, escalando y retornándolo a su posición original, se expresa como:

$$
(T_{-P_0})^{-1} \cdot S \cdot T_{-P_0} = \begin{pmatrix}
1 & 0 & 0 & x_0 \\
0 & 1 & 0 & y_0 \\
0 & 0 & 1 & z_0 \\
0 & 0 & 0 & 1
\end{pmatrix}^{-1}
\begin{pmatrix}
s_x & 0 & 0 & 0 \\
0 & s_y & 0 & 0 \\
0 & 0 & s_z & 0 \\
0 & 0 & 0 & 1
\end{pmatrix}
\begin{pmatrix}
1 & 0 & 0 & -x_0 \\
0 & 1 & 0 & -y_0 \\
0 & 0 & 1 & -z_0 \\
0 & 0 & 0 & 1
\end{pmatrix}
$$

Así, la operación de escalamiento en torno a un punto específico $P_0$ respecto al origen puede ser definida mediante la secuencia de transformaciones matriciales previamente indicada.

# Conclusión 


Esta actividad nos ha servido para comprender las diversas transformaciones tridiemnsionales revisadas en clase y aplicarlas en ejercicios prácticos. A través de la resolución de los problemas planteados, lo que nos sirve para aplicarlo a problemas más complejos en el futuro.

## Bibliografía

 - Xiang, Z., & Plastock, R. A. (2000). Schaum’s Outline of Computer Graphics 2/E. McGraw Hill Professional.